src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,334 @@
+  }
+
+  /**
+   * A function that parses the permissions object to check whether the specific principal has the specific permission types or not
+   *
+   * @param {Array} permissionTypes permission types
+   * @param {Object} principals the users or groups
+   * @returns true if the principal has the specified permission types, false if the principal has no permission
+   *
+   * @public
+   */
+  public hasPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || permissionTypes.length === 0 || !this.permissions || !principals) {
+      return false;
+    }
+
+    const currentPermissions = this.permissions;
+    return permissionTypes.some((permissionType) =>
+      checkPermission(currentPermissions[permissionType], principals)
+    );
+  }
+
+  /**
+   * A permissions object build function that adds principal with specific permission to the object
+   *
+   * This function is used to contruct a new permissions object or add principals with specified permissions to
+   * the existing permissions object. The usage is:
+   *
+   * const permissionObject = new ACL()
+   *  .addPermission(['write', 'management'], {
+   *     users: ['user2'],
+   *   })
+   *   .addPermission(['write', 'management'], {
+   *     groups: ['group1'],
+   *   })
+   *   .getPermissions();;
+   *
+   * @param {Array} permissionTypes the permission types
+   * @param {Object} principals the users or groups
+   * @returns the permissions object
+   *
+   * @public
+   */
+  public addPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || !principals) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      this.permissions[permissionType] = addToPrincipals({
+        principals: this.permissions[permissionType],
+        users: principals.users,
+        groups: principals.groups,
+      });
+    }
+
+    return this;
+  }
+
+  /**
+   * A permissions object build function that removes specific permission of specific principal from the object
+   *
+   * This function is used to remove principals with specified permissions to
+   * the existing permissions object. The usage is:
+   *
+   * const newPermissionObject = new ACL()
+   *  .removePermission(['write', 'management'], {
+   *     users: ['user2'],
+   *   })
+   *   .removePermission(['write', 'management'], {
+   *     groups: ['group1'],
+   *   })
+   *   .getPermissions();;
+   *
+   * @param {Array} permissionTypes the permission types
+   * @param {Object} principals the users or groups
+   * @returns the permissions object
+   *
+   * @public
+   */
+  public removePermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || !principals) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
```

Comment:
in the case if a user has `write` permission to an object, and we want to remove the (`read`) access to the object, how do we ensure the user principal is removed from the `write` list? is it controlled by the caller of this `removePermission` function?

Reply:
It's controlled by the caller, the caller needs to call `removePermission` and `addPermission` functions multiple times to construct a new `Permissions` object.